components/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/RoleManagementActionListener.java

```diff
@@ -127,4 +140,57 @@ private boolean isRoleSharingOperation() {
         String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
         return !StringUtils.equals(requestInitiatedOrganization, tenantDomain);
     }
+
+    /**
+     * Filters the list of user IDs to only include those that exist in the user store and are not from the agent
+     * identity userstore.
+     *
+     * @param userIds List of user IDs to filter.
+     * @return List of valid user IDs that exist in the user store and are not from the agent identity userstore.
+     * @throws IdentityRoleManagementException if an error occurs while checking user existence.
+     */
+    private List<String> filterAgents(List<String> userIds) throws IdentityRoleManagementException {
+
+        List<String> validUserIds = new ArrayList<>();
+        if (CollectionUtils.isEmpty(userIds)) {
+            return validUserIds;
+        }
+        AbstractUserStoreManager userStoreManager = getUserStoreManager();
```

Comment:
List<String> userNames = userStoreManager.getUserNamesFromUserIDs((List<String>) value);
// We can filter the username with `Agent` domain.

Reply:
getUserNamesFromUserIDs loops through getUserNameFromUserID, but I don’t see that it adds any value here.